ui/library/src/context/ScrollContext.ts

```diff
@@ -39,8 +64,121 @@ export function useScrollContextProps(): IScrollContext {
+    const set = new Set<number>();
+    Object.freeze(set);
+    return set;
+  });
+
+  const isOutlineTargetVisible = React.useCallback(
+    (dest: NodeDestination): boolean => {
+      return visibleOutlineTargets.has(dest);
+    },
+    [visibleOutlineTargets]
+  );
+
+  const isPageVisible = React.useCallback(
+    ({ pageNumber, pageIndex }: { pageNumber?: number; pageIndex?: number }): boolean => {
+      if (typeof pageIndex === 'number') {
+        pageNumber = pageIndex + 1;
+      }
+      if (typeof pageNumber !== 'number') {
+        return false;
+      }
+      return visiblePageNumbers.has(pageNumber);
+    },
+    [visiblePageNumbers]
+  );
+
+  const scrollToPage = React.useCallback(
+    ({ pageNumber, pageIndex }: { pageNumber?: number; pageIndex?: number }): void => {
+      if (typeof pageNumber === 'number') {
+        pageIndex = pageNumber - 1;
+      }
+      if (typeof pageIndex !== 'number') {
+        return;
+      }
+      document
+        .getElementById(generatePageIdFromIndex(pageIndex))
+        ?.scrollIntoView({ behavior: 'smooth' });
+    },
+    []
+  );
+
+  // Watch outline nodes
+  React.useEffect(() => {
+    const root = scrollRoot || document.documentElement;
+    const detector = new VisibilityDetector<NodeDestination>({
+      root: root,
+      setEntries: setVisibleOutlineNodes,
+      onVisibilityChange: ({ visibleElements, hiddenElements, lastEntries }) => {
+        const newEntries = new Set(lastEntries);
+        for (const el of hiddenElements) {
+          const dest = el.getAttribute('data-outline-target-dest');
+          newEntries.delete(dest);
+        }
+        for (const el of visibleElements) {
+          const dest = el.getAttribute('data-outline-target-dest');
+          newEntries.add(dest);
+        }
+        return newEntries;
+      },
+    });
+    detector.observeNodes('.reader__page__outline-target');
+    return () => {
+      detector.destroy();
+    };
+  }, [scrollRoot, observerIndex]);
+
+  // Watch pages
+  React.useEffect(() => {
+    const root = scrollRoot || document.documentElement;
+    const detector = new VisibilityDetector<number>({
+      root: root,
+      setEntries: setVisiblePageNumbers,
+      onVisibilityChange: ({ visibleElements, hiddenElements, lastEntries }) => {
+        const newEntries = new Set(lastEntries);
+        for (const el of hiddenElements) {
+          const elPage = el.querySelector('data-page-number');
+          const pageNumber = parseInt(elPage?.getAttribute('data-page-number') || '', 10);
+          newEntries.delete(pageNumber);
+        }
+        for (const el of visibleElements) {
+          const elPage = el.querySelector('data-page-number');
+          const pageNumber = parseInt(elPage?.getAttribute('data-page-number') || '', 10);
+          newEntries.add(pageNumber);
+        }
+        return newEntries;
+      },
+    });
+    detector.observeNodes('.reader__page');
```

Comment:
It would be great if you can create constants for this class name as well as the attribute names above so we can organize them easier.